File: 876-hand-of-straights/hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        if(hand.size() % groupSize != 0) return false;
        sort(hand.begin(),hand.end());
        map<int,int> res, need;
        for(auto ele: hand){
            res[ele]++;
            need[ele] = 0;
        }
        for(int i=0; i<hand.size(); i++){
            if(res[hand[i]] == 0 && need[hand[i]] > 0) return false;
            else if(need[hand[i]] == 0 && res[hand[i]] > 0){
                for(int j=hand[i]; j<hand[i]+groupSize; j++){
                    if(res.find(j) == res.end() || res[j] == 0) return false;
                    res[j]--;
                }
            } else {
                res[hand[i]]--;
                need[hand[i]]--;
            }
        }

        return true;
    }
};

File: 876-hand-of-straights/hand-of-straights.py
from collections import defaultdict
import heapq
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        
        # Check if the numbers can be grouped in groupSize
        if (len(hand) % groupSize) != 0:
            return False
        hashMap = defaultdict(int)
        
        for num in hand:
            hashMap[num] += 1
        
        minHeap = list(hashMap.keys())
        heapq.heapify(minHeap)
        
        while(len(minHeap) > 0):
            mini = minHeap[0]
            # if hashMap[mini] == 0:
            #     heapq.heappop(minHeap)
            #     continue
            
            for i in range(groupSize):
                if (mini+i) in hashMap:
                    hashMap[mini+i] -= 1
                    if hashMap[mini+i] == 0:
                        if (mini+i) != minHeap[0]:
                            return False
                        heapq.heappop(minHeap)
                else:
                    return False
        return True        